---
title: "common_gene_circular"
author: "Meng Li"
date: "2/2/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

setwd('D:/OneDrive - The Pennsylvania State University/Box Sync/Penn_State/Research_data/Lynch_Data_2020/eGWAS/MAGMA')

library(dplyr)
library(ggplot2)
library(tidyr)
library(tibble)
#library(pheatmap)
#  if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#  BiocManager::install("karyoploteR")
library(karyoploteR)
#library(qqman)
library(stringr)
library(data.table)
#BiocManager::install("ComplexHeatmap")
#library(ComplexHeatmap)
library(purrr)
library(circlize)
#library(gpart)
library(genetics)
library(ggplotify) # convert to grob
# ggplotify will take care of encapsulating the karyoploteR calls into grid or ggplot objects and make them compatible with grid, cowplot or patchwork. 
library(grid)
library(LDheatmap)
library(data.table)

```

## top gene super list

```{r gene}
# Data from MAGMA gene results: separate list
widiv_g <- read.csv('top_100_genes/widiv_g_top_100_genes_MAGMA_2.5kb_top_p.csv', row.names = 'X')
widiv_gxe <- read.csv('top_100_genes/widiv_gxe_top_100_genes_MAGMA_2.5kb_top_p.csv', row.names = 'X')
pred_root <- read.csv('top_100_genes/pred_root_top_100_genes_MAGMA_2.5kb_top_p.csv', row.names = 'X')
env <- read.csv('top_100_genes/env_top_100_genes_MAGMA_2.5kb_top_p.csv', row.names = 'X')

# top widiv and predroot genes
#top_genes_widiv_pred <- bind_rows(widiv_g = widiv_g, widiv_gxe = widiv_gxe, pred_root = pred_root, .id = 'group')
#write.csv(top_genes_widiv_pred,file = 'top_100_genes_widiv_g_gxe_predroot_MAGMA_2.5kb_long.csv')

# top widiv g and gxe genes
widiv <- bind_rows(widiv_g = widiv_g, widiv_gxe = widiv_gxe, .id = 'group') %>%
  group_by(GENE) %>% 
  summarise(CHR, START, STOP, trait = paste0(trait,collapse = ','), 
            group = paste0(unique(group),collapse = ','),P=max(P)) %>%
  unique() 

widiv <- widiv %>% 
  mutate(p.range = (P- min(widiv$P))/(max(widiv$P)-min(widiv$P))) %>%
  arrange(CHR,START)

env <- env %>%
  mutate(p.range = (P- min(env$P))/(max(env$P)-min(env$P))) %>%
  arrange(CHR,START)

pred_root <- pred_root %>%
  mutate(p.range = (P- min(pred_root$P))/(max(pred_root$P)-min(pred_root$P))) %>%
  arrange(CHR,START)

# top pred_root and env genes
#env_pred_root <- bind_rows(env = env, pred_root = pred_root, .id = 'group') %>%
#  group_by(GENE) %>% 
#  summarise(CHR, START, STOP, trait = paste0(trait,collapse = ','), 
#            group = paste0(unique(group),collapse = ','),P=max(P)) %>%
#  unique() %>% 
#  mutate(group_1 = case_when(group == 'env,pred_root' ~ 'overlap', TRUE ~ group))

# top env, pred_root, widiv
#env_pred_widiv <- bind_rows(env = env, pred_root = pred_root, widiv = widiv, .id = 'group') %>%
#  group_by(GENE) %>% 
#  summarise(CHR, START, STOP, trait = paste0(trait,collapse = ','), group = paste0(unique(group),collapse = ',')) %>%
#  unique() #%>% 
  #mutate(group_1 = case_when(group == 'env,pred_root' ~ 'overlap', TRUE ~ group))

# Data from MAGMA gene results: super list
top_gene_super <- read.csv('top_100_genes/top_100_genes_aggrate_from_4_lists_MAGMA_2.5kb.csv')

top_super_range <- top_gene_super %>% dplyr::select(-GENE)

# prepare data to convert to ranges
widiv_g.1 <- widiv_g %>% dplyr::select(-GENE) # Gene name is not needed
widiv_gxe.1 <- widiv_gxe %>% dplyr::select(-GENE) 
pred_root.1 <- pred_root %>% dplyr::select(-GENE)
env.1 <- env %>% dplyr::select(-GENE)
widiv.1 <- widiv[,-1]

```

## read root expression 

root expression data cleaning
```{r expression}
#super_exp <- read.csv('top_100_genes/top_100_genes_aggregate_from_4_lists_expression_2022-03-07.csv',  skip = 1)

#super_exp.1 <- super_exp %>%
#    mutate_all(~replace(., is.na(.), 0)) %>%
#    mutate(root_sum = rowSums(across(c(PR,Cortex,EZ,MZ,SR)))) %>% 
#    filter(!(root_sum == 0)) %>% # filter genes with no root expression data
#    dplyr::select(-c(chromosome,start, root_sum)) %>%
#    column_to_rownames(var = 'gene_name') %>%
#    mutate_all(~ log2(. +1)) %>% # log transformation log2(x+1)
#    rownames_to_column(var = 'GeneID') %>%
#    gather(key = 'tissue', value = 'value', -GeneID) %>%
#    group_by(GeneID) %>%
#    mutate(perc_95 = quantile(value, probs = 0.95)) %>% # calculate top 5% for each gene
#    spread(tissue, value) %>%
#    filter(PR >= perc_95 | Cortex >= perc_95| EZ >= perc_95 | MZ >= perc_95 | SR >= perc_95) %>%
#    mutate(perc_95_root = 1)

#super_exp_gene <- super_exp.1 %>% 
#  dplyr::select(GeneID) %>%
#  mutate(root_expression_95 = 1) %>%
#  dplyr::rename(GENE = GeneID)

# extract all high root expression gene (root expression > expressions in other tissues)
gene_expression <- read.csv('Gene_expression/all_gene_expression_avg_fpkm_Walley_2016.csv')

root_exp_all <- gene_expression %>%
    mutate_all(~replace(., is.na(.), 0)) %>% # convert na to zero
    mutate(root_sum = rowSums(across(c(PR,Cortex,EZ,MZ,SR)))) %>% 
    filter(!(root_sum == 0)) %>% # filter genes with no root expression data
    dplyr::select(-root_sum) %>%
    column_to_rownames(var = 'GENE') %>%
    mutate_all(~ log2(. +1)) %>% # log transformation log2(x+1)
    rownames_to_column(var = 'GENE') %>%
    pivot_longer(values_to = 'value', cols = -c(GENE, PR, Cortex,  EZ,  MZ, SR)) %>%
                   #endosperm_12DAP:Vegetative_Meristem_and_surrounding_Tissue )
    #gather(key = 'tissue', value = 'value', -GENE) %>%
    group_by(GENE) %>%
    #mutate(perc_95 = quantile(value, probs = 0.95)) %>% # calculate top 5% for each gene
    mutate(exp_max = max(value)) %>%
    #spread(tissue, value) %>%
    filter(PR > exp_max | Cortex > exp_max| EZ > exp_max | MZ > exp_max | SR > exp_max) %>%
    dplyr::select(-c(name,value)) %>%
    unique() %>%
    mutate(exp_max_root = 1)
  

write.csv(root_exp_all,'Gene_expression/root_expressed_genes_all.csv')

```

## combine all info

add informatiton from other datasets

```{r gene}
#all_info<-read.csv('common_gene/Super_list/common_genes_aggregate_from_4_lists_all_info.csv',header = TRUE, row.names = 'X')

## read root genes identified in other studies 
#RGPDB root genes
RGP <- read.csv('D:/OneDrive - The Pennsylvania State University/Box Sync/Penn_State/Research_data/Lynch_Data_2020/eGWAS/Overlap/Root_genes/Maize_master_RGPDB.csv')
RGP.1 <- RGP %>% dplyr::select(GeneID) %>% unique() %>% mutate(RGP = 1) %>% dplyr::rename(GENE = GeneID)

# Schneider root genes
SH_anatomy <- read.csv('D:/OneDrive - The Pennsylvania State University/Box Sync/Penn_State/Research_data/Lynch_Data_2020/eGWAS/Overlap/Root_genes/Root anatomy genes_Schneider_2020_Plant_Genome.csv') 
SH_anatomy.1 <- SH_anatomy %>% 
    dplyr::select(Group, trait, GeneID) %>% 
    group_by(GeneID) %>% 
    summarise(Group_SH = paste0(unique(Group),collapse = ','), trait_SH = paste0(unique(trait),collapse = ',')) %>% 
    dplyr::rename(GENE = GeneID) %>%
    unique() %>% 
    mutate(SH_root = 1) 
#SH_arch <- read.csv('D:/OneDrive - The Pennsylvania State University/Box Sync/Penn_State/Research_data/Lynch_Data_2020/eGWAS/Overlap/Root_genes/Root architecture_genes_Schneider_2020_J EXP BIOL.csv')
#SH_arch.1 <- SH_arch %>% 
#    dplyr::select(Group, trait, GeneID) %>% 
#    group_by(GeneID) %>% 
#    summarise(Group_SH = paste0(unique(Group),collapse = ','), trait_SH = paste0(unique(trait),collapse = ',')) %>% 
#    dplyr::rename(GENE = GeneID)

#SH_root <- rbind.data.frame(SH_anatomy.1,SH_arch.1)
#SH_root.1 <- SH_root %>% unique() %>% 
#    mutate(SH_root = 1, Group_SH = paste(Group_SH), trait_SH = paste(trait_SH)) 

# Kein metaxylem genes 
SK_mx <- read.csv('D:/OneDrive - The Pennsylvania State University/Box Sync/Penn_State/Research_data/Lynch_Data_2020/eGWAS/Overlap/Root_genes/Metaxylem_genes_Kein_2020_bioRxiv.csv')
SK_mx.1 <- SK_mx %>% 
    dplyr::select(Group, Phene, GeneID) %>%
    group_by(GeneID) %>% 
    summarise(Group_mx = paste0(unique(Group),collapse = ','), trait_mx = paste0(unique(Phene),collapse = ',')) %>%
    mutate(SK_mx = 1, Group_mx = paste(Group_mx), trait_mx = paste(trait_mx)) %>% 
    dplyr::rename(GENE = GeneID)

# GWAS atlas gene database
GWAS_atlas <- read.csv('D:/OneDrive - The Pennsylvania State University/Box Sync/Penn_State/Research_data/Lynch_Data_2020/eGWAS/Overlap/Root_genes/GWAS atlas_database_access_09172021.csv', fileEncoding="UTF-8-BOM")

GWAS_atlas.1 <- GWAS_atlas %>% 
  dplyr::rename(GENE=GeneID) %>%
  mutate(GWAS_atlas_root = case_when(str_detect(Reported_Traits, 'Root|root') ~ 1 )) %>%
  dplyr::select(GENE, Reported_Traits, GWAS_atlas_root )

# read GWAS atlas from MaizeGDB gff file
library(ape)
library(data.table)
GWAS_maizegdb <- read.gff('D:/OneDrive - The Pennsylvania State University/Box Sync/Penn_State/Research_data/Lynch_Data_2020/eGWAS/Overlap/Root_genes/GWAS_Atlas_db_MaizeGDB_access_09172021.gff3')

GWAS_maizegdb_1 <- GWAS_maizegdb %>% 
  separate(seqid, into = c('text','Chr'), sep = "(?<=[A-Za-z])(?=[0-9])") %>%
  filter(Chr %in% 1:10) %>%
  dplyr::select(-c(text, source,type,score, strand, phase)) 

GWAS_maizegdb_1$Chr <- as.numeric(GWAS_maizegdb_1$Chr)

# prepare data for str_remove
x <- data.table(GWAS_maizegdb_1)
y <- data.table(all_info)

setkey(y, CHR, START, STOP) # order y by Chr, Start, End

op <- foverlaps(x, y, by.x = c('Chr','start','end'),by.y = c('CHR','START','STOP'), type = 'any',nomatch = NULL)

GWAS_maizegdb.2 <- op %>% drop_na(GENE) %>% dplyr::select(GENE, attributes) %>%
  separate(attributes, into = c('Name', 'ID', 'FirstAuthor','Reference','Position','Tissue','Title',
                                'Chr','pvalue','PubMedID','RS','Alternate_allele','R2','Journal'), sep = ';') %>%
  mutate(Name = str_replace_all(Name,'%20',' '), FirstAuthor = str_replace_all(FirstAuthor,'%20',' '),
         Title = str_replace_all(Title, '%20',' '), Journal = str_replace_all(Journal, '%20',' ')) %>%
  mutate(Name = str_remove(Name, 'Name='), Tissue = str_remove(Tissue, 'Tissue='),
         FirstAuthor = str_remove(FirstAuthor, 'FirstAuthor='),
         Title = str_remove(Title, 'Title='),Journal = str_remove(Journal, 'Journal=')) %>%
  dplyr::select(GENE, Name, Tissue, FirstAuthor, Title, Journal) %>%
  group_by(GENE) %>%
  summarise(Name = paste(unique(Name), collapse = '; '),
            Tissue=paste(unique(Tissue), collapse = '; '), 
            FirstAuthor = paste(unique(FirstAuthor), collapse = '; '), 
            Title = paste(unique(Title), collapse = '; '),
            Journal = paste(unique(Journal), collapse = '; ')) %>%
      unique() %>%
  mutate(GWAS_maizegdb_root = case_when(str_detect(Tissue, 'Root|root')| str_detect(Name, 'Root|root')  ~ 1  )) %>%
  dplyr::select(GENE, Name,Tissue, GWAS_maizegdb_root )

# add gene symbol
Dan_anno <- read.csv("D:/OneDrive - The Pennsylvania State University/Box Sync/Penn_State/Research_data/Lynch_Data_2020/eGWAS/GATES_SITE_ANNOTATION/DAN_SNP_GENES_ANNO.csv")

Dan_anno.1 <- Dan_anno %>% 
  dplyr::select(GeneID, symbol.plus,Araport11, uniprot, pannzer2) %>%
  group_by(GeneID) %>%
  summarise(symbol.plus = paste(unique(symbol.plus), collapse = '; '),
            Araport11 = paste(unique(Araport11), collapse = '; '),
            uniprot = paste(unique(uniprot), collapse = '; '),
            pannzer2 = paste(unique(pannzer2), collapse = '; ')) %>%
  dplyr::rename(GENE = GeneID) %>%
  dplyr::select(GENE,Araport11)

# genes from maizegdb
gene_all <- fread(file = 'Gene_symbol_maizegdb/genes_all_downloaded_maizegdb_03082022.txt',fill=FALSE,na.strings= '')

gene_all.1 <- gene_all %>% 
  dplyr::select(`v4 Gene Model ID`, `Gene Symbol`, `Full Name`,`Source`) %>%
  drop_na(`v4 Gene Model ID`,`Full Name`) %>%
  unique() %>%
  group_by(`v4 Gene Model ID`) %>%
  summarise(maizegdb_symbol = paste(`Gene Symbol`, collapse = ','),
            Full_name = paste(`Full Name`, collapse = ','),
            Source = paste(Source, collapse = ','))

colnames(gene_all.1)[1] <- 'GENE' 

# genes related to root in Araport11 and was characterized in maize; these genes were scanned manually
gene_root <- fread(file = 'top_100_genes/filtered_root_genes_maize_symbol_araport11.csv',fill=FALSE,na.strings= '.')

#------combine------#
# aggregate information
all_info <- top_gene_super %>%
  dplyr::select(!c(group,trait)) %>%
  left_join(env[,1:5], by = c('GENE','CHR','START','STOP')) %>% 
  dplyr::rename(env = trait) %>%
  left_join(pred_root[,1:5], by = c('GENE','CHR','START','STOP')) %>%
  dplyr::rename(pred_root = trait) %>%
  left_join(widiv_g[,1:5], by = c('GENE','CHR','START','STOP')) %>%
  dplyr::rename(widiv_g = trait) %>%
  left_join(widiv_gxe[,1:5], by = c('GENE','CHR','START','STOP')) %>%
  dplyr::rename(widiv_gxe = trait) %>%
  #left_join(overlap_gene.variable,by = 'GENE') %>%
  left_join(root_exp_all[,c(1,8)], by = 'GENE') %>%
  arrange(CHR, START)

all_info.1 <- all_info %>% 
    left_join(RGP.1, by = 'GENE') %>%
    left_join(SH_anatomy.1, by = 'GENE') %>%
    left_join(SK_mx.1, by = 'GENE') %>%
    left_join(GWAS_atlas.1, by = 'GENE') %>%
    left_join(GWAS_maizegdb.2, by = 'GENE') %>%
    left_join(Dan_anno.1, by = 'GENE') %>%
    left_join(gene_all.1, by = 'GENE') %>%
    left_join(gene_root[,c(3,5,8,9)], by = 'GENE')

write.csv(all_info.1,'top_100_genes/top_100_genes_aggregate_from_4_lists_all_info_other_datasets.csv', row.names = FALSE,na = '.')

```

## set regions as +/- 200kb of the pred root genes

add info to regions

```{r region}
pred_root_region <- pred_root %>%
  arrange(CHR, START) %>%
  mutate(region_start = START - 200000, region_stop = STOP + 200000) %>% # add +/- 200kb
  dplyr::select(CHR,region_start,region_stop)

pred_root_region$CHR <- as.character(pred_root_region$CHR)

pred_root_region_rf_l <- rainfallTransform(pred_root_region, mode = 'left')
pred_root_region_rf_l$CHR <- as.integer(pred_root_region_rf_l$CHR)

pred_root_region_rf_l.1 <- pred_root_region_rf_l %>%
  arrange(CHR, region_start) %>% 
  replace_na(list(dist =1)) %>% # replcae na (chr breaks) with 200001
  mutate(region = cumsum((dist -1)>=0)) %>% # everytime the distance >200000, assign a new group number  >0
  group_by(region) %>%
  summarise(CHR,region_start = min(region_start), region_stop = max(region_stop)) %>%
  unique() 

# convert root region to a GRanges object
pred_root_region_GRanges <- makeGRangesFromDataFrame(pred_root_region_rf_l.1,keep.extra.columns = TRUE,ignore.strand = TRUE,seqnames.field = 'CHR', start.field = 'region_start', end.field = 'region_stop')

# convert all info dataframe to a GRanges object
all_info_GRanges <- makeGRangesFromDataFrame(all_info.1, seqnames.field = 'CHR',keep.extra.columns = TRUE, ignore.strand = TRUE,start.field = 'START',end.field = 'STOP')

# extract Genes with at least one marker overlap: notice the order query and subject
pred_root_region_overlap <- as.data.frame(findOverlaps(all_info_GRanges,pred_root_region_GRanges, ignore.strand = TRUE))

# extract genes that hit regions
pred_root_region_overlap.1 <- as.data.frame(all_info_GRanges[pred_root_region_overlap$queryHits]) %>%
  dplyr::rename(gene_start = start, gene_end = end, gene_width = width) %>%
  dplyr::select(-strand)

# extract region
pred_root_region_overlap.2 <- as.data.frame(pred_root_region_GRanges[pred_root_region_overlap$subjectHits]) %>%
  dplyr::rename(CHR = seqnames, region_start = start, region_end = end, region_width = width) %>%
  dplyr::select(region, CHR,region_start, region_end, region_width) 

#combine
pred_root_region_info <- cbind.data.frame(
  pred_root_region_overlap.2[,c('region','CHR','region_start','region_end','region_width')], pred_root_region_overlap.1[,-1])

write.csv(pred_root_region_info, file = 'top_100_genes/Region/all_regions_200kb_around_top_100_predroot_gene_all_info.csv',
          row.names = FALSE,na = '.')

```

a look at the regions
```{r region_info}
# how many regions contain widiv genes
n_widiv_region <- pred_root_region_info %>% 
  filter(!is.na(widiv_g) | !is.na(widiv_gxe)) %>% # 155 genes
  dplyr::select(region) %>%
  unique() # n = 89 regions

# how many regions contain env genes
n_env_region <- pred_root_region_info %>% 
  filter(!is.na(env)) %>% # 296 genes
  dplyr::select(region) %>%
  unique() # n = 160 regions

# how many regions contain both widiv and env genes
n_widiv_env_region <- intersect(n_env_region$region,n_widiv_region$region) # n = 33 regions

```

## filter regions:

selected region: at least one gene was found in previous published paper 
```{r select_region}
pred_root_region_info <- read.csv('top_100_genes/Region/all_regions_200kb_around_top_100_predroot_gene_all_info.csv',header=TRUE)
#if there is at least one gene was found in other datasets
pred_region_number_filter <- pred_root_region_info  %>%
  mutate(region_select = case_when(RGP == 1 | SH_root ==1 | SK_mx ==1 |GWAS_atlas_root == 1 | GWAS_maizegdb_root == 1 | Araport11_root ==1 | maize_symbol_root == 1 ~ 1)) %>%
  filter(region_select ==1) %>%
  dplyr::select(region) %>%
  unique()

pred_region_info.2 <- pred_root_region_info %>% # 230 regions
  filter(region %in% pred_region_number_filter$region) 

write.csv(pred_region_info.2, file ='top_100_genes/Region/filtered_gene_regions_root_genes_sources_around_pred_root_200kb.csv',na='.', row.names = FALSE)

```

a look at the filtered regions: 
```{r region_info}
# how many filtered regions contain widiv genes
n_widiv_filtered_region <- pred_region_info.2 %>% 
  filter(!is.na(widiv_g) | !is.na(widiv_gxe)) %>% # 41 genes
  dplyr::select(region) %>%
  unique() # n = 25 regions

# how many filtered regions contain env genes
n_env_filtered_region <- pred_region_info.2 %>% 
  filter(!is.na(env)) %>% # 84 genes
  dplyr::select(region) %>%
  unique() # n = 36 regions

# how many filtered regions contain both widiv and env genes
n_widiv_env_filtered_region <- intersect(n_env_filtered_region$region,n_widiv_filtered_region$region) # n = 9 regions

# how many filtered regions contain widiv or env root genes
n_pred_only_filtered_region <- pred_region_info.2 %>% 
  filter(!is.na(widiv_g) | !is.na(widiv_gxe) | !is.na(env)) %>% 
  dplyr::select(region) %>%
  unique() # n = 52 regions

```

cimmyt: extract selected regions used pred root: (Final:216,235, 404,417)
```{r extract}
#pred_region_info.4 <- read.csv(file ='top_100_genes/filtered_by_symbol_top_gene_regions_around_pred_root_200kb.csv',na.strings = '.')

pred_region_info.2 <- read.csv(file='top_100_genes/Region/filtered_gene_regions_root_genes_sources_around_pred_root_200kb.csv', header = TRUE)

select_region <- pred_region_info.2 %>% filter(region %in% c(216,235,404,417))

#select_region <- pred_region_info.2 %>% filter(region == 235)

select_region.1 <- select_region %>% 
    dplyr::select(region, CHR, region_start, region_end) %>%
    unique()
```

## load cimmyt genotypes, annotation

1. load cimmyt genotypes, annotation

```{r geno}
load('Cimmyt_magma_input/cimmyt_genotype_file.Rimage')
colm[colm ==3] <- NA

cimmyt_ann_all <- read.csv('./gene_annotation/all_gene_annotation_with_cimmyt_2.5kb_long.csv', header = TRUE,row.names = 'X')

```

add SNPs to selected region and gene: cimmyt (from all gene annotation info)

```{r SNP}
# all snps in the region
datalist <- list()
for(i in 1:nrow(select_region.1)) {
    SNP = cimmyt_ann_all %>% 
        filter(Chr == select_region.1$CHR[i] & (Physical.Position >= select_region.1$region_start[i] & Physical.Position <= select_region.1$region_end[i])) %>%
        mutate(region = select_region.1$region[i],
               region_start = select_region.1$region_start[i],
               region_end = select_region.1$region_end[i] ) %>%
        dplyr::select(region,region_start, region_end, Chr,GeneID, Start, End, Site, Physical.Position)
    
    datalist[[i]] <- SNP
}

cimmyt_snp <- dplyr::bind_rows(datalist) %>% unique()

cimmyt_snp_geno <- colm[intersect(row.names(colm),cimmyt_snp$Site),] %>% rownames_to_column(var = 'Site')
#cimmyt_snp_geno.1 <- cimmyt_snp_geno %>% column_to_rownames(var = 'Site') 
#cimmyt_snp_geno.1 <- as.data.frame(t(cimmyt_snp_geno.1))

#save(cimmyt_snp_geno.1, file = 'top_100_genes/cimmyt_select_region_snp_numeric_geno.Rimage')

# snps for selected genes
# genes of interest
# Region 404: Zm00001d047786; Region 417: Zm00001d023378, Zm00001d023379; Region 235: Zm00001d015397
# Region 216: Zm00001d012963, Zm00001d012964, Zm00001d012965
gene_select_404 <- cimmyt_snp %>% filter(GeneID %in% c('Zm00001d047786'))
snp_geno_404 <- colm[intersect(row.names(colm),gene_select_404$Site),] %>% rownames_to_column(var = 'Site')
# 235
gene_select_235 <- cimmyt_snp %>% filter(GeneID %in% c('Zm00001d015397'))
snp_geno_235 <- colm[intersect(row.names(colm),gene_select_235$Site),] %>% rownames_to_column(var = 'Site')
# 216
gene_select_216 <- cimmyt_snp %>% filter(GeneID %in% c('Zm00001d012963'))
snp_geno_216 <- colm[intersect(row.names(colm),gene_select_216$Site),] %>% rownames_to_column(var = 'Site')

#417: snps for 'Zm00001d023378','Zm00001d023379' were further filtered based on the position of the insertion
gene_select_417 <- cimmyt_snp %>% filter(GeneID %in% c('Zm00001d023378','Zm00001d023379'))
gene_select_417_Zm00001d023378 <- cimmyt_snp %>% filter(GeneID %in% c('Zm00001d023378'))
snp_geno_417 <- colm[intersect(row.names(colm),gene_select_417$Site),] %>% rownames_to_column(var = 'Site')
snp_geno_417_Zm00001d023378<- colm[intersect(row.names(colm),gene_select_417_Zm00001d023378$Site),] %>% rownames_to_column(var = 'Site')

rm(datalist,colm, SNP)
```

## load widiv genotypes, annotation

```{r geno}
library(data.table)

widiv <- fread('../GxE_GWAS/Widiv/widiv_2015_175g_MAF_0.05_Tassel_numeric.txt')

widiv_ann_all <- read.csv('./gene_annotation/all_gene_annotation_with_widiv_2.5kb_long.csv', header = TRUE,row.names = 'X')
```

add SNPs to selected region: widiv

```{r SNP}
datalist <- list()
for(i in 1:nrow(select_region.1)) {
    SNP = widiv_ann_all %>% 
        filter(Chr==select_region.1$CHR[i] & (BP>= select_region.1$region_start[i] & BP<= select_region.1$region_end[i])) %>%
        mutate(region = select_region.1$region[i],
               region_start = select_region.1$region_start[i],
               region_end = select_region.1$region_end[i] ) %>%
        dplyr::select(region,region_start, region_end, Chr,GeneID, Start, End, SNP, BP)
    
    datalist[[i]] <- SNP
}

widiv_snp <- dplyr::bind_rows(datalist) %>% unique()

widiv_snp_geno <- widiv %>% 
  dplyr::select(`<Marker>`, intersect(colnames(widiv),unique(widiv_snp$SNP))) %>%
  dplyr::rename(Genotype =`<Marker>` ) %>%
  column_to_rownames(var = 'Genotype') %>%
  t()
widiv_snp_geno<- as.data.frame(widiv_snp_geno) %>% rownames_to_column(var = 'SNP')

#widiv_snp_geno.1 <- widiv_snp_geno %>% column_to_rownames(var = 'SNP') 
#widiv_snp_geno.1 <- as.data.frame(t(widiv_snp_geno.1))

#save(widiv_snp_geno.1, file = 'top_100_genes/widiv_select_region_snp_numeric_geno.Rimage')

rm(datalist, widiv,SNP)

```

## Prepare data for miami plot

```{r heatmap}
#----widiv----#
widiv_names <- names(widiv_snp_geno)[1:176]

# create data object for LDheatmap
select_region_widiv <- widiv_snp %>%
    left_join(widiv_snp_geno, by = 'SNP') %>%
    group_by(region,Chr, region_start, region_end) %>%
    #mutate(SNP_n = n()) %>%
    #filter(SNP_n >1) %>% # filter out regions with only one SNP
    #dplyr::select(-SNP_n) %>%
    nest(BP = c(SNP,BP),GeneID = c(GeneID,Start,End), data = any_of(widiv_names)) 

# extract P values for miami plot
datalist.2 <- list()

for(i in 1:nrow(select_region_widiv)) {
    
    gene = bind_cols(region = select_region_widiv$region[[i]],
                     Chr = select_region_widiv$Chr[[i]],select_region_widiv$GeneID[[i]]) %>%
      dplyr::select(region,Chr,GeneID, Start, End) %>%
      dplyr::rename(CHR = Chr, GENE = GeneID, START = Start, STOP = End) %>%
      unique()
    
    datalist.2[[i]] <- gene
} 

widiv_genes <- dplyr::bind_rows(datalist.2)

#----cimmyt----#

cimmyt_names <- names(cimmyt_snp_geno)[1:1810]

# create data object for LDheatmap
select_region_cimmyt <- cimmyt_snp %>%
    left_join(cimmyt_snp_geno, by = 'Site') %>%
    group_by(region,Chr, region_start, region_end) %>%
    #mutate(SNP_n = n()) %>%
    #filter(SNP_n >1) %>% # filter out regions with only one SNP
    #dplyr::select(-SNP_n) %>%
    nest(BP = c(Site,Physical.Position),GeneID = c(GeneID,Start,End), data = any_of(cimmyt_names)) 

# extract P values for miami plot
datalist.2 <- list()

for(i in 1:nrow(select_region_cimmyt)) {
    
    gene = bind_cols(region = select_region_cimmyt$region[[i]],
                     Chr = select_region_cimmyt$Chr[[i]],select_region_cimmyt$GeneID[[i]]) %>%
      dplyr::select(region,Chr,GeneID, Start, End) %>%
      dplyr::rename(CHR = Chr, GENE = GeneID, START = Start, STOP = End) %>%
      unique()
    
    datalist.2[[i]] <- gene
} 

cimmyt_genes <- dplyr::bind_rows(datalist.2)

rm(datalist.2)

```

prepare data for miami plot
```{r miami}
#----widiv-----#
# Data from MAGMA gene results 
trait.list <- c('AA','RCA','MEDMETVA', 'MEDMETVD', 'TMETVA', 'NOMETV', 'TSA.RXSA','TSA.TCA')

# extract site names and combine into a single dataframe
combine_function <- function(x) { 
      
      df <-fread(file = paste0('Widiv_2.5kb_g/',x,'_G_GxE_GWAS_fitted_result_multi_g_2.5kb.genes.out.txt'), skip = 1)  # read text files, skip first line  # read text files, skip first line
      df_gxe <- fread(file= paste0('./Widiv_2.5kb_gxe/',x,'_G_GxE_GWAS_fitted_result_multi_gxe_2.5kb.genes.out.txt'), skip =1)
      
      #create containers for each datatype
      root_g <- df %>%
        dplyr::select(GENE, CHR,START,STOP,P_MULTI) %>%
        filter(GENE %in% widiv_genes$GENE) %>%
        mutate(trait = paste(str_split(x, fixed("_"))[[1]][1]), group = 'widiv_g') 
      
      root_gxe <- df_gxe %>%
        dplyr::select(GENE, CHR,START,STOP,P_MULTI) %>%
        filter(GENE %in% widiv_genes$GENE) %>%
        mutate(trait = paste(str_split(x, fixed("_"))[[1]][1]), group = 'widiv_gxe')
      
      root <- bind_rows(root_g,root_gxe)
      
}

data_list.test<-lapply(trait.list, combine_function)

# row bind dataframes
widiv_region_genes <-do.call(rbind,data_list.test) %>%
  mutate(trait = case_when(trait == 'MEDMETVA' ~ 'MMA',
                              trait == 'MEDMETVD' ~ 'MMD',
                              trait == 'TMETVA' ~ 'MVA',
                              trait == 'NOMETV' ~ 'NMV', TRUE ~ trait))
rm(data_list.test)

#----cimmyt pred root -----#
# Data from MAGMA gene results 
trait.list_pred <- c('AA','X.A','MMA', 'MMD', 'MVA', 'NMV', 'TSA.RXSA','TSA.TCA')

# extract site names and combine into a single dataframe
combine_function <- function(x) { 
      
      df_root <-fread(file = paste0('Predicted_roots/cimmyt_pred_mexi_',x,'_pred_pscovar_1656_MAF0.05.genes.out.txt'), skip = 1)
      #df_env <- fread(file = paste0('Cimmyt_env/cimmyt_mexi_',x,'_pscovar_1656_MAF0.05.genes.out.txt'), skip = 1)
        
      #create containers for each datatype
      root <- df_root %>%
        dplyr::select(GENE, CHR,START,STOP,P_MULTI) %>%
        filter(GENE %in% cimmyt_genes$GENE) %>%
        mutate(trait = paste(x), group = 'pred_root') 
      
      #env <- df_env %>%
      #  dplyr::select(GENE, CHR,START,STOP,P_MULTI) %>%
      #  filter(GENE %in% cimmyt_genes$GENE) %>%
      #  mutate(trait = paste(x), group = 'env')
      
      #cimmyt <- bind_rows(root,env)
      
}

data_list.pred_root<-lapply(trait.list_pred, combine_function)

# row bind dataframes
cimmyt_root_region_genes <-do.call(rbind,data_list.pred_root) %>%
  mutate(trait = case_when(trait == 'X.A' ~ 'RCA',TRUE ~ trait)) 

rm(data_list.pred_root)

#----cimmyt env ----#
env_list <- unique(c('CACO3', 'BS', 'PARWin', #AA
              'BS', 'PARFall', #X.A
              'coarse_1m', 'tmax_2', 'tmean_3', 'pet_4',  #MMA
              'tmax_2', 'tmax_3', 'tmax_4', 'tmin_3','tmean_3', 'tmean_5', 'Ann.Mean.Tmp', # MMD
              'tmax_1', 'tmax_2', 'tmax_3', 'tmax_4', 'tmax_11', 'tmax_12', 'tmin_3', 'tmin_4', 'tmin_5', 'tmean_2', 'tmean_3', 'tmean_4', 'tmean_5', 'tmean_11', 'pet_7', 'pet_8',  #MVA
              'tmax_10', 'tmax_11', 'tmin_3', 'tmin_6', 'tmin_8', 'tmean_5', 'tmean_10', 'tmean_11', 'Mean.Tmp.Dry.Q', 'pet_1', 'pet_12', # NMV
              'prec_3', 'prec_7', 'prec_10', 'prec_11', 'Ann.Prc', 'Prc.Wrm.Q', #TSA.RXSA
              'prec_2', 'prec_3', 'prec_7', 'prec_11', 'Ann.Prc', 'Prc.Wet.M', 'Prc.Wet.Q', 'Prc.Wrm.Q')) # TSA.TCA

# extract site names and combine into a single dataframe
combine_function <- function(x) { 
      
      df_env <- fread(file = paste0('Cimmyt_env/cimmyt_mexi_',x,'_pscovar_1656_MAF0.05.genes.out.txt'), skip = 1)
        
      env <- df_env %>%
        dplyr::select(GENE, CHR,START,STOP,P_MULTI) %>%
        filter(GENE %in% cimmyt_genes$GENE) %>%
        mutate(trait = paste(x), group = 'env')
      
}

data_list.env<-lapply(env_list, combine_function)

# row bind dataframes
cimmyt_env_region_genes <-do.call(rbind,data_list.env) 
rm(data_list.env)

# combine 
cimmyt_region_genes <- bind_rows(cimmyt_root_region_genes,cimmyt_env_region_genes)

```

extract top genes to highlight in miami plot
```{r highlight}
highlight_cimmyt <- select_region %>%
  dplyr::select(region, gene_start,gene_end,GENE, env, pred_root) 

# add NMV and tmax_3 info to highlighted data
#highlight_cimmyt <- rbind.data.frame(highlight_cimmyt,
#                                     c(235,88303760,88309464,'Zm00001d015397','.','NMV'),
#                                     c(235,88303760,88309464,'Zm00001d015397','tmax_3','.'))
    
highlight_cimmyt<- highlight_cimmyt %>%
  pivot_longer(values_to = 'trait',names_to = 'group', cols = env:pred_root,values_drop_na = TRUE) %>%
  separate(trait,into = as.character(c(1:20)), sep = ',') %>%
  pivot_longer(values_to = 'trait', values_drop_na = TRUE,cols = as.character(c(1:20))) %>%
  dplyr::select(-name) %>%
  left_join(cimmyt_region_genes, by =c('GENE','group','trait'))

highlight_widiv <- select_region %>%
  dplyr::select(region, gene_start,gene_end,GENE, widiv_g, widiv_gxe) %>%
  pivot_longer(values_to = 'trait',names_to = 'group', cols = widiv_g:widiv_gxe,values_drop_na = TRUE) %>%
  separate(trait,into = as.character(c(1:2)), sep = ',') %>%
  pivot_longer(values_to = 'trait', values_drop_na = TRUE,cols = as.character(c(1:2))) %>%
  dplyr::select(-name) %>%
  left_join(widiv_region_genes, by =c('GENE','group','trait'))

highlight_symbol <- select_region %>%
  mutate(root = case_when(GENE %in% c(#'Zm00001d027919','Zm00001d027925', # region 8
                                      #'Zm00001d028797', # region 15
                                      #'Zm00001d040702', # region 139
                                      #'Zm00001d052340', # region 196
                                      'Zm00001d012963', 'Zm00001d012964', 'Zm00001d012965', # region 216
                                      'Zm00001d015397', # region 235
                                      'Zm00001d047786', # region 404
                                      'Zm00001d023378', 'Zm00001d023379') ~ 1)) %>% # region 417
  drop_na(root)

```

## plot v29 gene on Chr 5
```{r v29}
# gene region information
gene_widiv = bind_cols(Chr = select_region_widiv$Chr,select_region_widiv$GeneID) %>%
    dplyr::select(Chr,GeneID, Start, End) %>%
    dplyr::rename(CHR = Chr, GENE = GeneID, START = Start, STOP = End) %>%
    unique()

gene_range_widiv <- toGRanges(as.data.frame(gene_widiv[,c(1,3,4)]))

gene_cimmyt = bind_cols(Chr = select_region_cimmyt$Chr,select_region_cimmyt$GeneID) %>%
    dplyr::select(Chr,GeneID, Start, End) %>%
    dplyr::rename(CHR = Chr, GENE = GeneID, START = Start, STOP = End) %>%
    unique()

gene_range_cimmyt <- toGRanges(as.data.frame(gene_cimmyt[,c(1,3,4)]))

# region information
widiv_region_genes_15 <- widiv_region_genes %>% 
    filter(GENE %in% filter(widiv_genes, region == select_region_widiv$region)$GENE) 

cimmyt_root_region_genes_15 <- cimmyt_root_region_genes %>% 
    filter(GENE %in% filter(cimmyt_genes, region == select_region_cimmyt$region)$GENE) 

cimmyt_env_region_genes_15 <- cimmyt_env_region_genes %>% 
    filter(GENE %in% filter(cimmyt_genes, region == select_region_cimmyt$region)$GENE) 

cimmyt_region_genes_15 <- bind_rows(cimmyt_root_region_genes_15,cimmyt_env_region_genes_15)

# convert region information to Granges
widiv_15_range <- toGRanges(data.frame(chr=widiv_region_genes_15$CHR,
                                       start=widiv_region_genes_15$START,
                                       end=widiv_region_genes_15$STOP,
                                       trait= widiv_region_genes_15$trait,
                                       group = widiv_region_genes_15$group,
                                       Gene = widiv_region_genes_15$GENE,
                                       pval=widiv_region_genes_15$P_MULTI))

cimmyt_15_range <- toGRanges(data.frame(chr=cimmyt_region_genes_15$CHR,
                                        start=cimmyt_region_genes_15$START,
                                        end=cimmyt_region_genes_15$STOP,
                                        trait= cimmyt_region_genes_15$trait,
                                        group = cimmyt_region_genes_15$group,
                                        Gene = cimmyt_region_genes_15$GENE,
                                        pval=cimmyt_region_genes_15$P_MULTI))

# highlight gene information
highlight_cimmyt_env_15 <- filter(highlight_cimmyt,group == 'env' & region == select_region.1$region) %>% drop_na(CHR)

highlight_cimmyt_pred_15 <- filter(highlight_cimmyt,group == 'pred_root' & region == select_region.1$region) %>% drop_na(CHR)

highlight_widiv_15 <- filter(highlight_widiv,region == select_region.1$region) %>% drop_na(CHR)

highlight_cimmyt_env_15_range <- toGRanges(data.frame(chr=highlight_cimmyt_env_15$CHR,
                                                      start=highlight_cimmyt_env_15$START,
                                                      end=highlight_cimmyt_env_15$STOP,
                                                      trait= highlight_cimmyt_env_15$trait,
                                                      group = highlight_cimmyt_env_15$group,
                                                      Gene = highlight_cimmyt_env_15$GENE,
                                                      y=-log10(highlight_cimmyt_env_15$P_MULTI),
                                                      pval=highlight_cimmyt_env_15$P_MULTI))

highlight_cimmyt_pred_15_range <- toGRanges(data.frame(chr=highlight_cimmyt_pred_15$CHR,
                                                       start=highlight_cimmyt_pred_15$START,
                                                       end=highlight_cimmyt_pred_15$STOP,
                                                       trait= highlight_cimmyt_pred_15$trait,
                                                       group = highlight_cimmyt_pred_15$group,
                                                       Gene = highlight_cimmyt_pred_15$GENE,
                                                       y=-log10(highlight_cimmyt_pred_15$P_MULTI),
                                                       pval=highlight_cimmyt_pred_15$P_MULTI))

highlight_widiv_15_range <- toGRanges(data.frame(chr=highlight_widiv_15$CHR,
                                                 start=highlight_widiv_15$START,
                                                 end=highlight_widiv_15$STOP,
                                                 trait= highlight_widiv_15$trait,
                                                 group = highlight_widiv_15$group,
                                                 Gene = highlight_widiv_15$GENE,
                                                 y=-log10(highlight_widiv_15$P_MULTI),
                                                 pval=highlight_widiv_15$P_MULTI))


highlight_symbol_15 <- highlight_symbol %>%
    filter(region == select_region.1$region)

highlight_symbol_15_range <- toGRanges(data.frame(chr=highlight_symbol_15$CHR,
                                                  start=(highlight_symbol_15$gene_start + highlight_symbol_15$gene_end)/2,
                                                  end=(highlight_symbol_15$gene_start + highlight_symbol_15$gene_end)/2,
                                                  Gene = highlight_symbol_15$GENE,
                                                  symbol = highlight_symbol_15$maizegdb_symbol))

# genome inormation
genome <- bind_cols(Chr = select_region.1$CHR, region_start = select_region.1$region_start, region_end = select_region.1$region_end)

custom.genome <- toGRanges(data.frame(chr=genome$Chr, start=genome$region_start,end= genome$region_end))

#c('Environment','Pred. root','WIDP g','WIDP gxe'), type = "points", background = NULL,pch = 21,
#    legend_gp = gpar(fill = c('orange','forestgreen',"#FF000080", "#0000FF80", 

#----miami plot----#
pdf(file = paste0('top_100_genes/Plot/select_region_miami_S5_v29_',select_region.1$region,'_new2.pdf'), width = 6, height = 4)# for a 
pp<-getDefaultPlotParams(plot.type=4) # show default plot parameters
pp$topmargin <- 6
pp$rightmargin <- 0.03
pp$leftmargin <- 0.08
pp$bottommargin <- 30
pp$ideogramheight <- 0 # change it to a line

#miami_15<-as.grob(expression(
kp.4 <- plotKaryotype(genome = custom.genome,plot.type=4,plot.params = pp) #cytobands = custom.cytobands,
kpAddBaseNumbers(kp.4, tick.dist = 100000, minor.tick.dist = 50000,
                 add.units = TRUE, cex=1, tick.len = 3, digits = 1)

#--cimmyt on the top--# highlight top 100 genes 
kpAddLabels(kp.4, labels = "-log10(P)", srt=90, pos=3, side = 'left', r0=0, r1=1, cex=1,label.margin = 0.06)
kpAddLabels(kp.4, labels = "CIMMYT", srt=90, pos=3, side = 'right', 
            r0=0, r1=1.5, cex=1, label.margin = 0)
kp <- kpPlotManhattan(kp.4, data=cimmyt_15_range, r0=0.5, r1=1, 
                      suggestive.col="#FFFFFF03", genomewide.col = "#FFFFFF03", points.cex = 1,
                      points.col = colByCategory(cimmyt_region_genes_15$group,colors =c("#BEBEBECC","#BEBEBECC")))
#adjust color to transparent: adjustcolor( "orange", alpha.f = 0.2) "#FFA50033"
#adjustcolor( "forestgreen", alpha.f = 0.2)[1] "#228B2233"

kpPoints(kp, data = highlight_cimmyt_env_15_range, pch=19, cex=1, col="orange",ymax=kp$latest.plot$computed.values$ymax, r0=0.5,r1=1) # highlight sig env points
kpPlotMarkers(kp, data = highlight_cimmyt_env_15_range, r0=0.5,r1=1,labels=highlight_cimmyt_env_15$trait,
              text.orientation = "horizontal", #marker.parts = c(0, 0.9, 0.1),
              line.color = "#FFFFFF03", label.color = 'orange' #label.dist = 0.01, max.iter = 1000
)
kpPoints(kp, data = highlight_cimmyt_pred_15_range, pch=19, cex=1, col="forestgreen",ymax=kp$latest.plot$computed.values$ymax, r0=0.5,r1=1) # highlight sig pred_root points
kpPlotMarkers(kp, data = highlight_cimmyt_pred_15_range, r0=0.5,r1=0.9,labels=highlight_cimmyt_pred_15$trait,
              text.orientation = "horizontal", #marker.parts = c(0, 0.9, 0.1),
              line.color = "#FFFFFF03", label.color = 'forestgreen' #label.dist = 0.01, max.iter = 1000
)

kpAxis(kp, ymin=0, ymax=ceiling(kp$latest.plot$computed.values$ymax),cex = 1, r0=0.5) #ceiling: round to the nearest #, label.margin = -100
# add gene regions
#kpPlotRegions(kp.4, data=gene_range_cimmyt, col="darkblue", border=darker('black'), r0=0.98, r1=1)

#--widiv on the bottom--# highlight top 100 genes: pick top genes and convert to range format to highlight
kpAddLabels(kp, labels = "WIDP", srt=90, pos=3, side = 'right', 
            r0=0, r1=0.5, cex=1, label.margin = 0)
kp <- kpPlotManhattan(kp.4, data=widiv_15_range, r0=0.5, r1=0, 
                      suggestive.col="white", genomewide.col = "white", points.cex = 1,
                      points.col = "#BEBEBECC")
kpPoints(kp, data = highlight_widiv_15_range, pch=19, cex=1, col="#FF000080",ymax=kp$latest.plot$computed.values$ymax, r0=0.5,r1=0) # highlight points
kpPlotMarkers(kp, data = highlight_widiv_15_range, r0=0.5,r1=0,labels=highlight_widiv_15_range$trait,
              text.orientation = "horizontal", #marker.parts = c(0, 0.9, 0.1),
              line.color = "#FFFFFF03", label.color = '#FF000080' #label.dist = 0.01, max.iter = 1000
)
kpAxis(kp, ymin=0, ymax=ceiling(kp$latest.plot$computed.values$ymax),cex =1 , r0=0.5, r1=0)
kpAbline(kp, h=0.5,col="black", lwd=1.5)
# add gene regions
#kpPlotRegions(kp, data=gene_range_widiv, col="darkblue", border=darker("black"), r0=0, r1=0.02)

#---add gene names---#
#kpPlotMarkers(kp, data=highlight_symbol_15_range, labels=highlight_symbol_15_range$Gene, srt=45, y=0, label.margin= 0.02,ymax=kp$latest.plot$computed.values$ymax,line.color=NULL, r0=0.15, cex =0.5)

#kpRect(kp, data=highlight_symbol_15_range, y0=0.05, y1=0.95, col=NA, border="black", lwd=0.2)

#---add focus area---#
kpAbline(kp, v=88303760-4000,col="black", lwd=1.5, lty= 2, data.panel = 1) # start of the focused area
kpAbline(kp, v=88309464+4000, col="black", lwd=1.5, lty=2, data.panel = 1) # end of the focused area

#---add legend---#
#legend(x = 0.7, y =1.2, fill = c("orange", "forestgreen"), box.col = 'white',xpd=TRUE,
#       legend = c("Environment", "Pred. root"))

dev.off()

```

## allele frequency

```{r allele}
library(data.table)
# read taxa names
cimmyt_pop <- fread(file = 'Cimmyt_magma_input/population_structure_n_1656.txt')
cimmyt_taxa <- as.data.frame(cimmyt_pop$id)
names(cimmyt_taxa) <- 'SaID'

cimmyt_AccID <- fread(file = '../Pop_structure/Mexi_All_Phenos_names_n=1809_01172022.csv')
cimmyt_taxa_1 <- cimmyt_taxa %>% left_join(cimmyt_AccID, by = 'SaID') %>% dplyr::select(SaID, AccID)

# read cluster id
library(readxl)
cimmyt_cluster <- read_excel("D:/OneDrive - The Pennsylvania State University/Box Sync/Penn_State/Research_data/Lynch_Data_2020/eGWAS/Pop_structure/PAM_clustering/CIMMYT_Anatomy_Clusters_k7.xlsx")
names(cimmyt_cluster)[1] <- 'AccID'
cimmyt_cluster_2 <- cimmyt_taxa_1 %>% left_join(cimmyt_cluster, by = 'AccID') %>% drop_na(cluster)

# read env and pred_root data
cimmyt_pred <- fread(file = 'Cimmyt_magma_input/cimmyt_predicted_root_pheno.txt') 
names(cimmyt_pred)[2] <- 'SaID'

cimmyt_env.1 <- fread(file = 'Cimmyt_magma_input/cimmyt_selected_env_pheno.txt')
names(cimmyt_env.1)[2] <- 'SaID'
cimmyt_env.2 <- fread(file = 'Cimmyt_magma_input/cimmyt_selected_env_pheno_2.txt')
names(cimmyt_env.2)[2] <- 'SaID'

cimmyt_env <- cimmyt_env.1[,-1] %>% left_join(cimmyt_env.2, by = 'SaID') %>% left_join(cimmyt_AccID[,c('SaID','Altitude_M')], by = 'SaID')

# top env for pred traits
# elevation
# MMA: coarse_1m, tmax_2, tmean_3, pet_4
# MMD: tmax_2, tmax_3, tmax_4, tmin_3, tmean_3, tmean_5, Ann.Mean.Tmp
# NMV: tmax_10, tmax_11, tmin_3, tmin_6, tmin_8, tmean_5, tmean_10, tmean_11, Mean.Tmp.Dry.Q, pet_1, pet_12
# MVA: tmax_1, tmax_2, tmax_3, tmax_4, tmax_11, tmax_12, tmin_3, tmin_4, tmin_5, tmean_2, tmean_3, tmean_4, tmean_5, tmean_11, pet_7, pet_8

#-------region 235------#
# read SNPs in selected region: 235

# calculate MAF function
getMAF <- function(m) colMeans(m, na.rm = TRUE) / 2 #this is calculating major allele for 0,1,2 genotypes

snp_235 <- snp_geno_235 %>%
    column_to_rownames(var = 'Site') %>%
    t() %>%
    as.data.frame() %>%
    rownames_to_column(var = 'SaID') 

snp_235_cluster <- cimmyt_cluster_2 %>% left_join(snp_235, by = 'SaID')

# filter minor allele frequency < 5%
snp_235_cluster_MAF <- as.data.frame(1-getMAF(snp_235_cluster[,4:24])) %>%
    rownames_to_column(var = 'Site') 
colnames(snp_235_cluster_MAF)[2] <- 'MAF'
snp_235_cluster_MAF <- snp_235_cluster_MAF %>% filter(MAF >= 0.05)

snp_235_filter <- snp_235_cluster %>% dplyr::select('SaID', snp_235_cluster_MAF$Site)

env_root_235 <- cimmyt_cluster_2 %>% 
    #left_join(cimmyt_pred[,c('SaID','MVA_pred')], by = 'SaID') %>%
    left_join(cimmyt_pred[,c('SaID','NMV_pred')], by = 'SaID') %>%
    #pivot_longer(names_to = 'root',values_to = 'pred', cols = MMA_pred:NMV_pred) %>%
    dplyr::select(-AccID) %>%
    group_by(cluster) %>%
    summarise(group_mean = mean(NMV_pred)) 

env_select_235 <- c('Altitude_M','Ann.Mean.Tmp')
                       #'tmax_1', 'tmax_2', 'tmax_3', 'tmax_4', 'tmax_11', 'tmax_12', 'tmin_3', 'tmin_4', 'tmin_5', 'tmean_2', 'tmean_3', 'tmean_4', 'tmean_5', 'tmean_11', 'pet_7', 'pet_8'))

cimmyt_env_select_235 <- cimmyt_env %>%
    dplyr::select(SaID,all_of(env_select_235))

env_root.data_235 <- cimmyt_cluster_2 %>%
    left_join(cimmyt_env_select_235,by = 'SaID') %>%
    left_join(snp_235_filter, by = 'SaID') %>%
    dplyr::select(-AccID) %>%
    #column_to_rownames(var = 'SaID') %>%
    pivot_longer(names_to = 'env',values_to = 'value', cols = Altitude_M:Ann.Mean.Tmp) %>%
    group_by(cluster, env) %>%
    nest() %>%
    mutate(env_mean = map(data, ~ mean(.$value)),
        df1 = map(data, ~ as.data.frame(1-getMAF(.[,-c(1,20)])))) %>% # this is calculating minor allele frequency
    #dplyr::select(id, df1) %>%
    mutate(MAF = purrr::map(df1, ~ rownames_to_column(., var = "SNP"))) %>%
    dplyr::select(cluster,env, env_mean, MAF) %>%
    unnest(c(MAF, env_mean)) %>%
    left_join(env_root_235, by = 'cluster') %>%
    as.data.frame() 

names(env_root.data_235)[5] <- 'MAF'

# ggplot regression plot + correlation P ####
library(ggpubr)
plot_235 <-
  ggplot(data=env_root.data_235, aes(x =MAF)) + 
      geom_smooth(aes(y = env_mean),method = 'lm',color="#E66100", se=TRUE, fill = "#E66100") +
      geom_point(aes(y = env_mean), color = "brown3") +
      #stat_cor(aes(x=MAF, y =TN_mean),label.x = 0.1,label.y = 0.2) +
      stat_cor(aes(x=MAF, y = env_mean),label.x.npc = 0.4,label.y.npc = 0.1, color = 'brown3') +
      #geom_smooth(aes(y = (group_mean- 384.077)*1000), color = '#5D3A9B', method ='lm', se = TRUE,fill = '#5D3A9B') + #color = rgb(0.2, 0.6, 0.9, 1), 
      #geom_point( aes(y = (group_mean - 384.077)*1000), color = 'darkblue') + #,color = 'darkblue'
      facet_wrap(~ SNP + env, scales = 'free', ncol = 4) +
      #stat_cor(aes(x=MAF, y =(group_mean- 384.077)*1000),label.x.npc = 0.6,label.y.npc = 0.9,color = 'darkblue') +
      xlab('Minor allele frequency') +
      #labs(title = 'ASV37_bacteria_rhizosphere_LN_pred') +
      #scale_y_continuous(limits = c(0, 1),breaks = c(0,0.5,1), expand = c(0,0), name = 'Total nitrogen (%)', #
      #                   sec.axis = sec_axis(~.*0.001+384.077 , name = 'ASV abundance')) + #
      theme_bw() +
        theme(
        axis.title.y = element_text( size=13),
        #axis.title.y.right = element_text(color = 'darkblue', size=13),
        panel.grid = element_blank()
  ) 

plot_235_root <-
  ggplot(data=filter(env_root.data_235, env == 'Altitude_M'), aes(x =MAF)) + 
      geom_smooth(aes(y = group_mean),method = 'lm',color="#E66100", se=TRUE, fill = "#E66100") +
      geom_point(aes(y = group_mean), color = "brown3") +
      #stat_cor(aes(x=MAF, y =TN_mean),label.x = 0.1,label.y = 0.2) +
      stat_cor(aes(x=MAF, y = group_mean),label.x.npc = 0.4,label.y.npc = 0.1, color = 'brown3') +
      #geom_smooth(aes(y = (group_mean- 384.077)*1000), color = '#5D3A9B', method ='lm', se = TRUE,fill = '#5D3A9B') + #color = rgb(0.2, 0.6, 0.9, 1), 
      #geom_point( aes(y = (group_mean - 384.077)*1000), color = 'darkblue') + #,color = 'darkblue'
      facet_wrap(~ SNP, scales = 'free', ncol = 4) +
      #stat_cor(aes(x=MAF, y =(group_mean- 384.077)*1000),label.x.npc = 0.6,label.y.npc = 0.9,color = 'darkblue') +
      xlab('Minor allele frequency') +
      ylab('NMV_pred') +
      #labs(title = 'ASV37_bacteria_rhizosphere_LN_pred') +
      #scale_y_continuous(limits = c(0, 1),breaks = c(0,0.5,1), expand = c(0,0), name = 'Total nitrogen (%)', #
      #                   sec.axis = sec_axis(~.*0.001+384.077 , name = 'ASV abundance')) + #
      theme_bw() +
        theme(
        axis.title.y = element_text( size=13),
        #axis.title.y.right = element_text(color = 'darkblue', size=13),
        panel.grid = element_blank()
  ) 
plot_235_root


```

## focused snps: maps, allele frequency

```{r snps}
# ggplot map
#library(raster)
library(rworldmap)
library(ggplot2)
library(ggh4x) # for changing panel size

map.polygon <- getMap(resolution = "low") # country border

# 235 snp: S5_88306826, S5_88306863
map_235_data <- snp_235_cluster %>% 
    left_join(cimmyt_AccID, by = c('AccID','SaID'))

map_235<- ggplot() +
    geom_path(data = map.polygon, aes(x = long, y = lat, group = group)) + # world map
    geom_point(data=map_235_data,aes(x = Longitude, y = Latitude, color = factor(S5_88306826)),size = 0.5) +
    scale_color_manual(values = c('red','purple','blue'), name='S5_88306826') +
    coord_sf(xlim = c(-120, -82), ylim = c(12, 34), expand = FALSE) + #focused area
    force_panelsizes(rows = unit(4, "cm"),cols = unit(6, "cm")) + # set the panel size
    theme_bw() +
    theme(axis.ticks = element_blank(),
          axis.text.y = element_blank(),
          axis.text.x = element_blank(),
          #axis.title.y = element_text(size =4, vjust = 0.5,color = 'black',face = 'bold', family = 'sans'),
          axis.title = element_blank(),
          #axis.title.x = element_text(size =4, vjust = 0.5,color = 'black',face = 'bold', family = 'sans'),
          legend.title = element_text(size = 5),
          plot.margin = unit(c(0,0,0,0), "cm"),
          panel.grid.major = element_blank(),
          panel.background = element_rect(fill = 'white'))

map_235

# save as a pdf
ggsave(plot = map_235, filename = 'top_100_genes/Plot/MAF/region_235_S5_88306826_map_6x4cm.pdf', height=6, width=8, units='cm', dpi=300)


#------------- MAF regression map for selected snps----------------#
library(ggpubr)

cluster.palette <- c("#66C2A5","#FC8D62","#8DA0CB","#E78AC3","#A6D854","#FFD92F","#E5C494")

# 235 snp: S5_88306826, S5_88306863
plot_235_elevation <-
  ggplot(data=filter(env_root.data_235, env == 'Altitude_M' & SNP == 'S5_88306863' ), aes(x =MAF)) + #ASV == 'ASV37_bacteria_rhizosphere_CK_pred'&
      geom_smooth(aes(y = env_mean),method = 'lm',color="lightgrey", se=TRUE, fill = "lightgrey") +
      geom_point(aes(y = env_mean, color = factor(cluster)), size = 4) +
      scale_color_manual(values = cluster.palette, name = 'Cluster') +
      #scale_x_continuous(limits = c(0,0.35), breaks = c(0, 0.1, 0.2, 0.3), expand = c(0,0)) +
      stat_cor(aes(x=MAF, y = env_mean),label.x.npc = 0.4,label.y.npc = 0.1, color = 'black') +
      xlab('Minor allele frequency') +
      ylab('Elevation (m)') +
      theme_bw() +
        theme(
        axis.title = element_text( size=13),
        #axis.title.y.right = element_text(color = 'darkblue', size=13),
        axis.line = element_line(color = 'black'),
        axis.text = element_text(color = 'black'),
        panel.grid = element_blank()
  ) 
plot_235_elevation

# save as svg: 600*400

plot_235_root <-
      ggplot(data=filter(env_root.data_235, env == 'Altitude_M' & SNP == 'S5_88306863' ), aes(x =MAF)) + #ASV == 'ASV37_bacteria_rhizosphere_CK_pred'&
      geom_smooth(aes(y = group_mean),method = 'lm',color="darkgrey", se=TRUE, fill = "darkgrey") +
      geom_point(aes(y = group_mean, color = factor(cluster)), size = 4) +
      scale_color_manual(values = cluster.palette, name = 'Cluster') +
      #scale_x_continuous(limits = c(0,0.35), breaks = c(0, 0.1, 0.2, 0.3), expand = c(0,0)) +
      #stat_cor(aes(x=MAF, y =TN_mean),label.x = 0.1,label.y = 0.2) +
      stat_cor(aes(x=MAF, y = group_mean),label.x.npc = 0.1,label.y.npc = 0.1, color = 'black') +
      #geom_smooth(aes(y = (group_mean- 384.077)*1000), color = '#5D3A9B', method ='lm', se = TRUE,fill = '#5D3A9B') + #color = rgb(0.2, 0.6, 0.9, 1), 
      #geom_point( aes(y = (group_mean - 384.077)*1000), color = 'darkblue') + #,color = 'darkblue'
      #facet_wrap(~ root, scales = 'free', ncol = 2) +
      #stat_cor(aes(x=MAF, y =(group_mean- 384.077)*1000),label.x.npc = 0.6,label.y.npc = 0.9,color = 'darkblue') +
      xlab('Minor allele frequency') +
      ylab('NMV') +
      #labs(title = 'ASV37_bacteria_rhizosphere_LN_pred') +
      #scale_y_continuous(limits = c(0, 1),breaks = c(0,0.5,1), expand = c(0,0), name = 'Total nitrogen (%)', #
      #                   sec.axis = sec_axis(~.*0.001+384.077 , name = 'ASV abundance')) + #
      theme_bw() +
        theme(
        axis.title = element_text( size=13),
        #axis.title.y.right = element_text(color = 'darkblue', size=13),
        axis.line = element_line(color = 'black'),
        axis.text = element_text(color = 'black'),
        panel.grid = element_blank()
  ) 

plot_235_root

```